tests/common.py

```diff
@@ -160,3 +160,60 @@ def mock_pending_merge_repo_paths(repo_name, src=True, pending=True):
         os.makedirs(path.parent, exist_ok=True)
         with open(path, "w") as fd:
             fd.write(PENDING_MERGE_FILE_TMPL.format(repo_name=repo_name, pid="1234"))
+
+
+class MockCompletedProcess:
+    def __init__(self, args=None, stdout=None, returncode=0):
+        self.args = args
+        self.returncode = returncode
+        self.stdout = stdout
+
+
+def mock_subprocess_run(mock_spec=None):
+    """Return a function that you can pass to unittest.mock.patch:
+
+    with patch("subprocess.run", mock_subprocess_run(spec)):
+        do your test
+
+    mock_speck is a list of dictionaries. The entries are used to match
+    subsequent calls to subprocess.run(), in order.
+
+    Each directory has the following keys:
+
+    args: (required) the arguments that subprocess.run() is expected to be called with.
+    If this is a callable, then it is run with the args received and is expected to test them and return True
+
+    stdout: (optional) a string that will be used as the output of the command
+
+    sim_call: (optional) a callable that will be executed instead of the process
+
+    sim_call_args: (optional) a list of positional parameters to `sim_call`
+
+    sim_call_kwargs: (optional) a list of named parameters to `sim_call`
+
+    The mock returns a MockCompletedProcess object. If stdout is not None, the string or bytes object
+    passed is is the `stdout` attribute of that object.
+    """
+    if mock_spec is None:
+        mock_spec = []
+
+    def mock_run(args, stdout=None, **kwargs):
+        call_spec = mock_spec.pop(0)
+        if call_spec["args"] is not None:
```

Comment:
wouldn't be more clear to have the mock yield a an obj to test? Eg:

```
with patch("subprocess.run", mock_subprocess_run) as mocked:
    # run your stuff
    mocked.assert_args(...)
    mocked.assert_sim_call(...)
```
It's just a bit hard to understand what is tested in the tests below just because everything is wrapped into the mocker.

However, I bet you had your reasons to go this way...

Reply:
Not sure if I get you. You mean memorizing the args passed to the mocker  and checking them after the whole thing ran ? Feasible, but then I don't think I can run my sim_call functions to produce the side effects (esp. creating a couple of files which are expected by further steps of the process, or producing output). 

I'll think about it.